### DISCUSSION TITLE
help building ecal with recent protobuf version

### DISCUSSION THREAD

**Discussion by ENAC-Robotique** [eclipse-ecal/ecal#1602] [Q&A]
Hi,
I'm trying to build ecal with a recent protobuf version (e.g. v26.1) or a RaspberryPi4 and I am facing some issues.

- First of all, I tried setting `ECAL_THIRDPARTY_BUILD_PROTOBUF=ON` and setting `Protobuf_VERSION` to 26.1, but I could not get it working.
- I decided to build (and install) protobuf independently first:
    - I cloned protobuf and try building and installing it. protoc was complaining about absl.
    - I had to clone and build from sources Abseil to make protobuf working
    - I build and install protobuf again (v26.1) and made it works
- then I resumed the ecal build, with `ECAL_THIRDPARTY_BUILD_PROTOBUF=OFF`, and I have this error:
```
/usr/bin/ld: /usr/local/lib/libprotobuf.so: undefined reference to symbol 'AbslInternalSpinLockWake_lts_20240116'
/usr/bin/ld: /usr/local/lib/libabsl_spinlock_wait.so.2401.0.0: error adding symbols: DSO missing from command line
```

This is surprising because it seems that the symbol is present in the shared library:
```
$ nm /usr/local/lib/libprotobuf.so | grep AbslInternalSpinLockWake_lts_20240116
                 U AbslInternalSpinLockWake_lts_20240116
```

Do you have an idea on how I can solve this issue ?

**Comment by KerstinKeller** [comment]
Hi,

so first of all, which Release / Branch of eCAL are you building?
We recommend using either 5.13.1 or 5.12.5 (we've had some minor bug reports for 5.13.1 but working on it).

Regarding ECAL_THIRDPARTY_BUILD_PROTOBUF=ON: when this is set, protobuf will be build from submodule, which ist fixed to 3.11.4 and without any abseil dependency, so its to be expected that this approach doesn't work.

On Windows, I have successfully build eCAL (5.12.5, 5.13.x) with Protobuf 3.26.1, however I build both Protobuf and Abseil as static libraries, using the most recent Abseil release.
Did you build both Abseil and Protobuf as .so?
I found [this stackoverflow post](https://stackoverflow.com/questions/19901934/libpthread-so-0-error-adding-symbols-dso-missing-from-command-line) suggesting that its a matter of arguments (or even argument order) specified to the linker.

Maybe something is wrong in the CMakeLists.txt / ProtobufConfig.cmake files, and how Abseil is linked when eCAL links to `protobuf::libprotobuf`.
What's the call with all arguments to your linker?

**Reply by Fabien-B** [reply]
Hi, same person here.
I am on the `support/v5.13` branch, 2 commits above `v5.13.1`. (`204764`)

I was indeed building protobuf as a shared library, so I tried again as static library:
- I built Abseil `20240116.2`. I made sure that static libraries were installed.
- I build protobuf `v26.1` with `protobuf_ABSL_PROVIDER=package` to use the Abseil I just built. I made sure static libraries were installed
- I tried building ecal again. I have an other error this time:

```
ld: [...]/libabsl[...].a: relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol [...] which may bind externally can not be used when making a shared object; recompile with -fPIC
```

I read somewhere that I should replace `find_package(Protobuf REQUIRED)` by `find_package(protobuf CONFIG REQUIRED)` on recent protobuf versions so I tried that, without any success.

I am using gcc 13.2.0 on a fresh Ubuntu 24.04.

**Reply by ENAC-Robotique** [reply]
I resolved the relocation problem by building abseil and protobuf with `-DCMAKE_POSITION_INDEPENDENT_CODE=ON` but now I'm back to the undefined reference:
```
/usr/bin/ld: ../../../lib/libecal_core.so.5.13.1: undefined reference to `bool absl::lts_20240116...
```

What is very curious is that I can find the symbols in `libecal_core.so` with `nm`...

**Reply by KerstinKeller** [reply]
Are you able to print the linker arguments when trying to link? I think we need this line when trying to figure out what's going wrong.
And how is the exact CMake invokation for eCAL when you're trying to build?
Maybe the full log of the build would help here.
